frontend/src/metabase/visualizations/echarts/mixins.ts

```diff
@@ -0,0 +1,201 @@
+import type { EChartsOption } from "echarts";
+
+import { VisualizationProps } from "../types";
+
+const MOCK_TOOLTIP_DATA = {
+  index: 3,
+  event: {
+    isTrusted: true,
+  },
+  stackedTooltipModel: {
+    headerTitle: "Total",
+    headerRows: [
+      {
+        name: "80  –  100",
+        value: 65,
+        color: "#EF8C8C",
+      },
+    ],
+    bodyRows: [
+      {
+        name: "20  –  40",
+        value: 94,
+        color: "#98D9D9",
+      },
+      {
+        name: "40  –  60",
+        value: 77,
+        color: "#F9D45C",
+      },
+      {
+        name: "60  –  80",
+        value: 52,
+        color: "#A989C5",
+      },
+      {
+        name: "Other",
+        value: 10,
+        color: "#949AAB",
+      },
+    ],
+    showTotal: true,
+    showPercentages: true,
+  },
+};
+
+type EChartsEventHandler = {
+  // TODO better types
+  eventName: string;
+  query?: string;
+  handler: (event: any) => void;
+};
+
+type ZREventHandler = {
+  // TODO better types
+  eventName: string;
+  handler: (event: any) => void;
+};
+
+export type EChartsConfig = {
+  option: EChartsOption;
+  eventHandlers: EChartsEventHandler[];
+  zrEventHandlers: ZREventHandler[];
+};
+
+type EChartsMixin = (params: {
+  chartType: any;
+  props: VisualizationProps;
+  option: EChartsOption;
+}) => {
+  option: EChartsOption;
+  eventHandlers?: EChartsEventHandler[];
+  zrEventHandlers?: ZREventHandler[];
+};
+
+export const lineSeriesMixin: EChartsMixin = ({ option }) => {
+  // TODO real implementation
+  return {
+    option: {
+      ...option,
+      xAxis: {
+        type: "category",
+        data: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
+      },
+      yAxis: {
+        type: "value",
+      },
+      series: [
+        {
+          name: "Some Series Name",
+          data: [150, 230, 224, 218, 135, 147, 260],
+          type: "line",
+        },
+      ],
+    },
+  };
+};
+
+export const smoothSettingMixin: EChartsMixin = ({
+  props: { settings },
+  option,
+}) => {
+  if (Array.isArray(option?.series)) {
+    option.series.forEach(series => {
+      series.smooth = settings.smooth;
```

Comment:
@alxnddr Do you think mutating the `option` object directly would cause any issues? We can always do a pure functional approach instead but I figured there's no harm in mutating it and doing so makes the code simpler

Reply:
I don't see any harm in mutations here as well but I wonder if we should not, at least for the very beginning, try split every viz setting in a separate mixin. I assume it would add more complexity to properly match and mutate the series slice on line/area/bar chart series settings. Let me think more about it